File: js/utils/state.js
/** 오목의 모든 규칙은 이 개체를 상속하여 제작해야함 */
class State{
    /** 초기화 
     * @param {Board} me 아군
     * @param {Board} enemy 적군
    */
    constructor(me, enemy){
        this.me = me;
        this.enemy = enemy;
        this.rule = new Rule(this.me, this.enemy);
    }

    /** 패배 여부 확인 
     * @return {boolean}
    */
    is_lose() {
        return this.enemy.is_win;
    }
    /** 무승부 여부 확인 */
    is_draw(){
        return this.me.count(1) + this.enemy.count(1) == this.me.size ** 2;
    }

    /** 게임 종료 여부를 확인 */
    is_done(){
        return this.is_lose() || this.is_draw();
    }

    /** 선수 여부 확인 
     * @return {boolean}
    */
    is_first_player(){
        return this.me.is_first;
    }

    // 하위 개체에서 무조건 구현해야할 함수들

    /** 다음상태로 전환하는 함수 
     * @param {number} action 어디로 둘지를 결정
    */
    next(action){
        throw new Error("이 함수는 무조건 구현해야하는 함수입니다.");
    }

    /** 이길수 있는 합법적인 수의 리스트를 얻기 
     * @param {number[]} forbid 금수의 위치를 담을 변수를 받음
    */
    smart_legal_actions(forbid=[]){
        throw new Error("이 함수는 무조건 구현해야하는 함수입니다.");
    }

    /** 합법적인 수의 리스트 얻기 
     * @param {number[]} cause 금수의 위치를 담을 변수를 받음
    */
    legal_actions(cause=[]){
        throw new Error("이 함수는 무조건 구현해야하는 함수입니다.");
    }

    /** 깊은 복사 */
    deepcopy(){
        throw new Error("이 함수는 무조건 구현해야하는 함수입니다.");
    }

    /** 리셋 */
    reset(){
        throw new Error("이 함수는 무조건 구현해야하는 함수입니다.");
    }
}

File: train/state.py
# 게임 상태
from abc import ABC, abstractmethod
from board import Board
from rule import Rule


class State(ABC):
    # 초기화
    def __init__(self, me:Board, enemy:Board):
        self.me = me
        self.enemy = enemy
        self.rule = Rule(self.me, self.enemy)

    # 패배 여부 확인
    def is_lose(self):
        return self.enemy.is_win

    # 무승부 여부 확인
    def is_draw(self):
        return self.me.count(1) + self.enemy.count(1) == self.me.size ** 2
        
    # 게임 종료 여부 확인
    def is_done(self):
        return self.is_lose() or self.is_draw()

    # 선수 여부 확인
    def is_first_player(self):
        return self.me.is_first

    # 문자열 표시
    def __str__(self):
        return (self.me + self.enemy).__str__()

    # 다음 상태 얻기
    @abstractmethod
    def next(self, action:int):
        """
        구현 목록
        - action에 따라 me 보드 상태 수정
        - 수정된 상태에 승자가 있는지 판별
        - 다음상태 리턴
        """
        pass

    # 이길 수 있는 합법적인 수의 리스트 얻기
    @abstractmethod
    def smart_legal_actions(self, forbid={}):
        """
        구현 목록
        - 이길 수 있는 action 리스트릴 리턴
        """
        pass

    # 합법적인 수의 리스트 얻기
    @abstractmethod
    def legal_actions(self, cause=[]):
        """
        구현 목록
        - 규칙에 따라 둘수 있는 action 리스트를 리턴
        """
        pass





if __name__ == "__main__":
    class A(State):
        pass

    try:
        a = A(None, None)
    except Exception as e:
        print(e)
        